libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,598 @@
+            } else {
+                AvoidanceHandler &tmp = handler_for_action((MAV_COLLISION_ACTION)action);
+                new_avoidance_handler = &tmp;
+            }
+
+            if (new_avoidance_handler != _current_avoidance_handler) {
+                GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_WARNING, "AVOID: Moving to avoidance handler (%s)", ((new_avoidance_handler != nullptr) ? new_avoidance_handler->name() : "NULL"));
+                if (new_avoidance_handler == nullptr) {
+                    if (_current_avoidance_handler != nullptr) {
+                        _current_avoidance_handler->exit();
+                        _current_avoidance_handler = nullptr;
+                    }
+                } else {
+                    if (new_avoidance_handler->enter(*threat, _current_avoidance_handler)) {
+                        _current_avoidance_handler = new_avoidance_handler;
+                    } else {
+                        // whinge. This may not be an internal error -
+                        // e.g. we may have lost GPS
+                    }
+                }
+            }
+        }
+        _old_state = new_state;
+    }
+
+    if (_current_avoidance_handler != nullptr) {
+        _current_avoidance_handler->update();
+    }
+}
+
+
+
+// thanks to grumdrig (http://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment)
+float length_squared(const Vector2f v, const Vector2f w)
+{
+    Vector2f tmp = v-w;
+    return tmp.length_squared();
+}
+float dot(const Vector2f v, const Vector2f w)
+{
+    return v*w;
+}
+float distance(const Vector2f v, const Vector2f w)
+{
+    Vector2f delta = v-w;
+    return delta.length();
+}
+
+float closest_distance_between_line_and_point(const Vector2f &v,
+                                              const Vector2f &w,
+                                              const Vector2f &p)
+{
+    // Return minimum distance between line segment vw and point p
+    const float l2 = length_squared(v, w);  // i.e. |w-v|^2 -  avoid a sqrt
+    if (l2 < FLT_EPSILON) return distance(p, v);   // v == w case
+    // Consider the line extending the segment, parameterized as v + t (w - v).
+    // We find projection of point p onto the line. 
+    // It falls where t = [(p-v) . (w-v)] / |w-v|^2
+    // We clamp t from [0,1] to handle points outside the segment vw.
+    const float t = MAX(0, MIN(1, dot(p - v, w - v) / l2));
+    const Vector2f projection = v + (w - v) * t;  // Projection falls on the segment
+    return distance(p, projection);
+}
+// end thanks to grumdrig (http://stackoverflow.com/questions/849211/shortest-distance-between-a-point-and-a-line-segment)
+
+
+// derived from above, assuming v is (0,0)
+float closest_distance_between_radial_and_point(const Vector2f &w,
+                                                const Vector2f &p)
+{
+    const float l2 = w.length_squared();
+    if (l2 < FLT_EPSILON) return p.length();
```

Comment:
This code was taken straight from the attributed code.

I was avoiding rewriting it, but I think it is annoying people too much with the different code style.  Hmmm.